src/cid_collections/frozen_vec.rs

```diff
@@ -0,0 +1,139 @@
+use super::*;
+use cid::Cid;
+use serde::{Deserialize, Serialize};
+
+#[cfg(doc)]
+use crate::blocks::TipsetKeys;
+
+/// There are typically MANY small, immutable collections of CIDs in, e.g [`TipsetKeys`].
+///
+/// Save space on those by:
+/// - Using a boxed slice to save on vector allocations.
+/// - Using [`SmallCid`]s
+///
+/// This may be expanded to have [`smallvec`](https://docs.rs/smallvec/1.11.0/smallvec/index.html)-style indirection
+/// to save more on heap allocations.
+#[derive(Default, Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct FrozenCidVec {
+    inner: Box<[SmallCid]>,
+}
+
+impl FrozenCidVec {
+    pub fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+    pub fn contains(&self, cid: Cid) -> bool {
+        self.inner.contains(&SmallCid::from(cid))
+    }
+}
+
+/// A [`MaybeCompactedCid`], with indirection to save space.
+///
+/// This is NOT a general purpose type - other collections should use the variants
+/// of [`MaybeCompactedCid`], so that the discriminant is not repeated.
+///
+/// (This saves ~20% in the [typical case](https://github.com/ChainSafe/forest/issues/3498#issuecomment-1723528854))
```

Comment:
This is either incorrect or misleading. There are a few different things at play here. Firstly, there's a gain in memory efficiency when going from plain `Cid` to `SmallCid`. Cids take up 96 bytes of memory, SmallCid takes 40 bytes. But using `CidHashSet` rather than `HashSet<SmallCid>` _also_ yields better efficiency at 32 bytes per entry rather than 40. This is the gain you measured in the link above.

Also, saying it saves ~20% unnecessarily complicates matters. The saving is exactly 25% for each CID. You measured ~24% because there are other allocated structures than just CIDs.